### PR TITLE
asynchronous-object-saving outside game loop.

### PR DESCRIPTION
**Pull request by Forbiddenz**
Moved DB saving into its own game and implemented a blocking collection model.
Card - asynchronous-object-saving

### CODE REVIEW

File: Source/ACE/Managers/DbManager.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            // starts game loop.
+            running = true;
+            new Thread(Tick).Start();
+        }
+
+        private static void Tick()
+        {
+            while (running)
+            {
+                Thread.Sleep(1);
```
Comment (ddevec) [inline Source/ACE/Managers/DbManager.cs]: The Thread.Sleep() is unneeded, as the BlockingCollection manages waiting.

**Comment by Forbiddenz** [comment]
Ok, let me try to add that in, today a bit crazy.  I also would like to try figure out a a better async reporting method maybe in here..  I am just a little wary of the await function inside this thread.

**Comment by Mogwai-TheFurry** [comment]
Also:
* Character saving on another thread is good.  Moving ALL saving to another thread is better.  Making characters but not items async could lead to bad race conditions and object duplication.
* There's no graceful shutdown mechanism.  The Tick loop needs to check a flag that can be set on the class for a graceful shutdown after all work is done.

File: Source/ACE.Database/ICommonDatabase.cs [hunk 2]
```diff
@@ -15,5 +15,6 @@ public interface ICommonDatabase
         AceObject GetObject(uint aceObjectId);
 
         Task<bool> SaveObject(AceObject aceObject);
+        Task<bool> SaveObjectAsync(AceObject aceObject);
```
Comment (Mogwai-TheFurry) [inline Source/ACE.Database/ICommonDatabase.cs]: Normally, you would have this:
```
bool SaveObject(AceObject aceObject);
Task<bool> SaveObjectAsync(AceObject aceObject);
```

File: Source/ACE.Database/IShardDatabase.cs [hunk 3]
```diff
@@ -34,5 +34,8 @@ public interface IShardDatabase : ICommonDatabase
         uint GetNextCharacterId();
 
         AceCharacter GetCharacter(uint id);
+
+        // call back error interfaces
+        void UnableToSaveObject(AceObject ao);
```
Comment (Mogwai-TheFurry) [inline Source/ACE.Database/IShardDatabase.cs]: This isn't part of the database interface - you need a callback to the calling code, not just some other function in your current class.

File: Source/ACE.Database/ShardDatabase.cs [hunk 4]
```diff
@@ -420,6 +420,35 @@ public uint RenameCharacter(string currentName, string newName)
...
+                UnableToSaveObject(aceObject);
+                result = false;
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Hooking into this interface will catch a saveobject error.
+        /// </summary>
+        public void UnableToSaveObject(AceObject ao)
```
Comment (Mogwai-TheFurry) [inline Source/ACE.Database/ShardDatabase.cs]: Take out the UnableToSaveObject stuff - your logging will do for now.

**Pull request merged**